pennylane/data/attributes/list.py

```diff
@@ -0,0 +1,124 @@
+from typing import Generic, List, Union, overload
+
+from pennylane.data.base.attribute import AttributeType
+from pennylane.data.base.mapper import MapperMixin
+from pennylane.data.base.typing_util import HDF5Any, HDF5Group, T
+
+
+class DatasetList(
+    Generic[T],
+    AttributeType[HDF5Group, typing.Sequence[T], typing.Iterable[T]],
+    MutableSequence,
+    MapperMixin,
+):
+    """Provides a list-like collection type for dataset attributes."""
+
+    type_id = "list"
+
+    def __post_init__(self, value: typing.Iterable[T], info):
+        self.extend(value)
+
+    def default_value(self) -> typing.Iterable[T]:
+        return ()
+
+    def hdf5_to_value(self, bind: HDF5Group) -> typing.MutableSequence[T]:
+        return self
+
+    def value_to_hdf5(
+        self, bind_parent: HDF5Group, key: str, value: typing.Iterable[T]
+    ) -> HDF5Group:
+        grp = bind_parent.create_group(key)
+
+        return grp
+
+    def copy_value(self) -> List[T]:
+        return [self._mapper[str(i)].copy_value() for i in range(len(self))]
+
+    def copy(self) -> List[T]:
+        """Returns a copy of this list as a builtin ``list``, with all
+        elements copied.."""
+        return self.copy_value()
+
+    def insert(self, index: int, value: Union[T, AttributeType[HDF5Any, T, T]]):
+        """Implements the insert() method."""
+        if index < 0:
+            index = len(self) + index
+
+        if index < 0:
+            index = 0
+        elif index >= len(self):
+            self._mapper[str(len(self))] = value
+            return
+
+        for i in reversed(range(index, len(self))):
+            self._mapper.move(str(i), str(i + 1))
+
+        self._mapper[str(index)] = value
+
+    def __len__(self) -> int:
+        return len(self.bind)
+
+    def __eq__(self, __value: object) -> bool:
+        if not isinstance(__value, Sequence):
+            return False
+
+        if not len(self) == len(__value):
+            return False
+
+        return all(x == y for x, y in zip(self, __value))
+
+    def __str__(self) -> str:
+        return str(list(self))
+
+    def __repr__(self) -> str:
+        return repr(list(self))
+
+    @overload
+    def __getitem__(self, index: slice) -> typing.List[T]:
+        ...
+
+    @overload
+    def __getitem__(self, index: int) -> T:
+        ...
+
+    def __getitem__(self, index: Union[int, slice]):
+        if isinstance(index, slice):
+            return [self[i] for i in range(len(self))[index]]
+
+        if index < 0:
+            index = len(self) + index
+
+        if index < 0 or index >= len(self):
+            raise IndexError(index)
+
+        return self._mapper[str(index)].get_value()
+
+    def __setitem__(self, index: int, value: Union[T, AttributeType[HDF5Any, T, T]]):
+        if index < 0:
+            index = len(self) + index
+
+        if index < 0 or index >= len(self):
+            raise IndexError(index)
+
+        self._mapper[str(index)] = value
```

Comment:
I'm not sure this works as intended. Check it out:
```python
>>> import pennylane as qml
>>> x = qml.data.attributes.DatasetList([1, 2, 3])
>>> x[1] = 4
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mikhail/ro/pennylane/pennylane/data/attributes/list.py", line 121, in __setitem__
    self._mapper[str(index)] = value
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/mapper.py", line 117, in __setitem__
    self.set_item(key, value, None)
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/mapper.py", line 112, in set_item
    attr_type(value, info, parent_and_key=(self.bind, key))
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/attribute.py", line 246, in __init__
    self._bind = self._set_value(value, info, parent, key)
  File "/Users/mikhail/ro/pennylane/pennylane/data/base/attribute.py", line 315, in _set_value
    new_bind = self.value_to_hdf5(parent, key, value)
  File "/Users/mikhail/ro/pennylane/pennylane/data/attributes/scalar.py", line 33, in value_to_hdf5
    bind_parent[key] = value
  File "/Users/mikhail/ro/pennylane/venv/lib/python3.9/site-packages/h5py/_hl/group.py", line 484, in __setitem__
    h5o.link(ds.id, self.id, name, lcpl=lcpl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5o.pyx", line 202, in [h5py.h5o.link](http://h5py.h5o.link/)
OSError: Unable to create link (name already exists)
```

Reply:
fixed